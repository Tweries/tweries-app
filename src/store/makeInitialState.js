import { NEWLINE } from '../constants';
import makeTweetstorm from './makeTweetstorm';

function makeInitialState({ feature }) {
  return {
    hashtags: '',
    healthy: false,
    items: makeTweetstorm(feature)({
      hashtags: '',
      linefeed: NEWLINE,
      source: ''
    }),
    linefeed: NEWLINE,
    notification: null,
    source: '',
    userId: ''
  };
}

export default makeInitialState;
