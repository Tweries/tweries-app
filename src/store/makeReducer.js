import { SUCCESS } from '../constants';
import makeInitialState from './makeInitialState';
import makeTweetstorm from './makeTweetstorm';

export const types = {
  APPEND_SCREEN_NAME: 'APPEND_SCREEN_NAME',
  CHANGE_SOURCE: 'CHANGE_SOURCE',
  CHANGE_TWEET: 'CHANGE_TWEET',
  DISMISS_TOAST: 'DISMISS_TOAST',
  RESET_TWEETSTORM: 'RESET_TWEETSTORM',
  SET_HEALTHY: 'SET_HEALTHY',
  SET_USER_ID: 'SET_USER_ID'
};

function makeReducer(feature) {
  function reducer(state = makeInitialState({ feature }), action) {
    switch (action.type) {
      case types.APPEND_SCREEN_NAME: {
        const source = `@${action.value} ${state.source}`;
        return {
          ...state,
          items: makeTweetstorm(feature)({
            linefeed: state.linefeed,
            source
          }),
          source
        };
      }
      case types.CHANGE_SOURCE: {
        return {
          ...state,
          items: makeTweetstorm(feature)({
            linefeed: state.linefeed,
            source: action.value
          }),
          source: action.value
        };
      }
      case types.CHANGE_TWEET: {
        return {
          ...state,
          items: [...state.items].map((item) => {
            const copy = { ...item };
            if (copy.id === action.value.id) {
              copy.tweet = action.value.tweet;
            }
            return copy;
          })
        };
      }
      case types.DISMISS_TOAST: {
        return {
          ...state,
          notification: null
        };
      }
      case types.RESET_TWEETSTORM: {
        if (action.value.type === SUCCESS) {
          return {
            ...makeInitialState({ feature }),
            healthy: state.healthy,
            notification: action.value,
            userId: state.userId
          };
        }
        return {
          ...state,
          notification: action.value
        };
      }
      case types.SET_HEALTHY: {
        return {
          ...state,
          healthy: action.value
        };
      }
      case types.SET_USER_ID: {
        return {
          ...state,
          userId: action.value
        };
      }
      default:
        return state;
    }
  }

  return reducer;
}

export default makeReducer;
