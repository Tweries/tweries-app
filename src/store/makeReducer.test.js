import makeReducer, { types } from './makeReducer';
import makeInitialState from './makeInitialState';

const USER_ID = 'twitter|1183836409850814464';

const feature = { active: () => true };
const base = { ...makeInitialState({ feature }) };
const notification = { message: 'Oh Noes!', type: 'danger' };

describe('reducer', () => {
  const scenarios = [
    {
      description: types.CHANGE_HASHTAGS,
      props: {
        action: {
          type: types.CHANGE_HASHTAGS,
          value: '#FOO'
        },
        state: { ...base, source: 'bar' }
      }
    },
    {
      description: types.CHANGE_LINEFEED,
      props: {
        action: {
          type: types.CHANGE_LINEFEED,
          value: '||'
        },
        state: {
          ...base,
          source: 'Can you can a can [..] as a canner || can can a can?'
        }
      }
    },
    {
      description: types.CHANGE_SOURCE,
      props: {
        action: {
          type: types.CHANGE_SOURCE,
          value: 'bar'
        },
        state: base
      }
    },
    {
      description: types.CHANGE_TWEET,
      props: {
        action: {
          type: types.CHANGE_TWEET,
          value: {
            id: 'baz',
            tweet: 'quux'
          }
        },
        state: {
          ...base,
          items: [{ id: 'baz', tweet: 'qux' }, { id: 'quuz', tweet: 'corge' }],
          source: 'qux'
        }
      }
    },
    {
      description: types.DISMISS_TOAST,
      props: {
        action: {
          type: types.DISMISS_TOAST
        },
        state: {
          ...base,
          notification
        }
      }
    },
    {
      description: types.RESET_TWEETSTORM,
      props: {
        action: {
          type: types.RESET_TWEETSTORM,
          value: notification
        },
        state: {
          ...base,
          hashtags: '#FOO',
          source: 'bar',
          healthy: true,
          userId: USER_ID
        }
      }
    },
    {
      description: types.SET_HEALTHY,
      props: {
        action: {
          type: types.SET_HEALTHY,
          value: true
        },
        state: base
      }
    },
    {
      description: types.SET_USER_ID,
      props: {
        action: {
          type: types.SET_USER_ID,
          value: USER_ID
        },
        state: base
      }
    },
    {
      description: 'default',
      props: {
        action: { type: 'UNKNOWN', value: undefined },
        state: undefined
      }
    }
  ];

  scenarios.forEach(({ description, props: { action, state } }) => {
    it(description, () => {
      const { items, ...rest } = makeReducer(feature)(state, action);
      expect(rest).toMatchSnapshot();
      items.forEach(item => {
        expect(item).toMatchSnapshot({
          id: expect.any(String)
        });
      });
    });
  });
});
