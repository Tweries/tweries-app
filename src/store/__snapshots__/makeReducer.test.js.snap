// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer APPEND_SCREEN_NAME 1`] = `
Object {
  "hashtags": "",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "@TweriesApp is awesome",
  "userId": "",
}
`;

exports[`reducer APPEND_SCREEN_NAME 2`] = `
Object {
  "id": Any<String>,
  "tweet": "@TweriesApp is awesome",
}
`;

exports[`reducer CHANGE_HASHTAGS 1`] = `
Object {
  "hashtags": "#FOO",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "bar",
  "userId": "",
}
`;

exports[`reducer CHANGE_HASHTAGS 2`] = `
Object {
  "id": Any<String>,
  "tweet": "bar #FOO",
}
`;

exports[`reducer CHANGE_SOURCE 1`] = `
Object {
  "hashtags": "",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "bar",
  "userId": "",
}
`;

exports[`reducer CHANGE_SOURCE 2`] = `
Object {
  "id": Any<String>,
  "tweet": "bar",
}
`;

exports[`reducer CHANGE_TWEET 1`] = `
Object {
  "hashtags": "",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "qux",
  "userId": "",
}
`;

exports[`reducer CHANGE_TWEET 2`] = `
Object {
  "id": Any<String>,
  "tweet": "quux",
}
`;

exports[`reducer CHANGE_TWEET 3`] = `
Object {
  "id": Any<String>,
  "tweet": "corge",
}
`;

exports[`reducer DISMISS_TOAST 1`] = `
Object {
  "hashtags": "",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "",
  "userId": "",
}
`;

exports[`reducer RESET_TWEETSTORM 1`] = `
Object {
  "hashtags": "",
  "healthy": true,
  "linefeed": "
",
  "notification": Object {
    "message": "Oh Noes!",
    "type": "danger",
  },
  "source": "",
  "userId": "twitter|1183836409850814464",
}
`;

exports[`reducer SET_HEALTHY 1`] = `
Object {
  "hashtags": "",
  "healthy": true,
  "linefeed": "
",
  "notification": null,
  "source": "",
  "userId": "",
}
`;

exports[`reducer SET_USER_ID 1`] = `
Object {
  "hashtags": "",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "",
  "userId": "twitter|1183836409850814464",
}
`;

exports[`reducer default 1`] = `
Object {
  "hashtags": "",
  "healthy": false,
  "linefeed": "
",
  "notification": null,
  "source": "",
  "userId": "",
}
`;
